core/course_controller/CourseController.py

```diff
@@ -1,48 +1,226 @@
+            if course_code is provided, it matches an existing course in the Courses table.
         Post-conditions: Returns the id of the modified Courses table object. throws an
-            IllegalValue exception if id is invalid or the courseData was invalid.
+            IllegalValue exception if course_code is invalid or the courseData was invalid.
         Side-effects: Inserts or updates a record in the Courses table with the given
             course_data
+        Returns: the id of the modified Courses table object. throws an
+            IllegalValue exception if course_code is invalid or the courseData was invalid.
         """
-        pass
+        if not course_data.course_code:
+            raise ValueError("Course code cannot be empty.")
 
+        # Handle course update
+        if course_code:
+            try:
+                course = Course.objects.get(course_code=course_code)
+            except Course.DoesNotExist:
+                raise ValueError("Course code does not match any existing course.")
+            course.course_name = course_data.course_name or course.course_name
+            course.semester = course_data.semester or course.semester
+            course.save()
+        else:
+            # Handle new course creation
+            if Course.objects.filter(course_code=course_data.course_code).exists():
+                raise ValueError("Course code already exists.")
+            if not course_data.semester:
+                raise ValueError("Semester is required for creating a new course.")
 
+            course = Course.objects.create(
+                course_code=course_data.course_code,
+                course_name=course_data.course_name,
+                semester=course_data.semester,
+            )
+
+        # Save lab sections
+        if course_data.lab_sections_codes:
+            CourseController.save_lab_sections(course_data)
+
+        # Save course sections
+        if course_data.course_sections_codes:
+            CourseController.save_course_sections(course_data)
+
+        return str(course.id)
 
     @staticmethod
-    def get_course(course_id: str) -> CourseOverview:
+    def get_course(course_code: str) -> CourseOverview:
         """
-        Pre-conditions: id provided is a valid course_id
+        Pre-conditions: course_code provided is a valid course_code
         Post-conditions: Returns an object containing course, sections, and assignment
-            information for object in the Courses table with the given id.
+            information for object in the Courses table with the given course_code.
         Side-effects: N/A
         """
-        pass
+        try:
+            course = Course.objects.get(course_code=course_code)
+        except Course.DoesNotExist:
+            raise ValueError("Course with the given code does not exist.")
+
+        default_section = CourseSection.objects.filter(course=course).first()
+        instructor = UserRef(
```

Comment:
we had talked about only having one instructor but to make this code "extensible" we should probably make it a list of instructors still and do it the right way at the lower level even if we don't show the full functionality at first in the view